### DISCUSSION TITLE
Best way to add an alias for a dimension name

### DISCUSSION THREAD

**Discussion by matthew-brett** [pydata/xarray#8570] [Q&A] [answered]
I hope the answer is not obvious, but I could not see it on a fairly quick search.

We have need of aliases for dimensions in our brain imaging datasets.

For example, for 4D MRI images, we might reasonably chose to label the dimensions as `i, j, k, time`.  But, any one of the `i, j, k` dimensions can also be the `slice` dimension (the scanner usually collects the images on 2D slice at a time).   So `slice` could be an alias for any of `i` or `j` or `k`.   Perhaps we could set this information with:

```python
img['k'].attrs['alias'] = 'slice'
```

We might want to index over the `slice` dimension, but this would involve us first - checking whether any dimension has a `slice` alias, and then indexing over the found dimension.  We would really like to be able to use syntax like `img.isel(slice=0)` - to mean the same - in this case - as `img.isel(k=0)`. 

Is there a canconical way of achieving that (using an alias for the standard dimension name)?   Or would we have to make a custom accessor to do this - e.g. img.our_accessor.isel(slice=0)`?

**Comment by dcherian** [comment]
You have three options AFAICT:
1. Create a duplicate `slice` variable
2. https://docs.xarray.dev/en/stable/generated/xarray.Dataset.filter_by_attrs.html
3. https://cf-xarray.readthedocs.io/en/latest/custom-criteria.html for an implementation of the custom accessor idea.

Xarray tries to not interpret `attrs` at all (since its hard to consistently propagate them, and its easy for them to be out of date). Exceptions are reading/writing data, or plotting.

**Reply by matthew-brett** [reply]
Thanks for the quick reply.  Could you say more about creating a duplicate `slice` variable?   Won't `img.isel(<name>=0)` need `<name>` to be a value in the `dims` tuple?

**Reply by dcherian** [reply]
This isn't well  documented unfortunately but you can do this
```python
img = xr.DataArray(dims=("i", "j", "k"), data=np.ones((120, 120, 120)), coords={"k": ("k", np.arange(120))})
img["slice"] = img["k"]
img = img.set_xindex("slice")
img.sel(slice=40)
```

**Reply by matthew-brett** [reply]
Thanks again!  But - I noticed that:

```python
img.isel(slice=40)
```

after your code, still fails with:

```
ValueError: Dimensions {'slice'} do not exist. Expected one or more of ('i', 'j', 'k')
```
Is this expected?

**Reply by dcherian** [reply, marked as answer]
Yes. `set_xindex` sets an "index" which translates from coordinate-space to index-space. We haven't enabled "index-space" selection with `.isel` on such additional coordinates. (cc @benbovy )

For completeness here's the `cf-xarray` version if you wanted to experiment with the accessor style interface:
```python
import xarray as xr
import numpy as np

img = xr.DataArray(
    dims=("i", "j", "k"),
    data=np.ones((120, 120, 120)),
    coords={"k": ("k", np.arange(120))},
)
img.k.attrs["alias"] = "slice"

import cf_xarray as cfxr  # registers .cf

cfxr.set_options(custom_criteria={"slice": {"alias": "slice"}})
img.cf.isel(slice=2)
img.cf.sel(slice=2)
```
